src/openssl.c

```diff
@@ -5487,6 +5488,85 @@ static int xr_setPublicKey(lua_State *L) {
 } /* xr_setPublicKey() */
 
 
+static int xr_setExtensionByNid(lua_State *L, X509_REQ *csr, int target_nid, void* value) {
+	STACK_OF(X509_EXTENSION) *sk = NULL;
+	X509_ATTRIBUTE *attr;
+	int has_attrs=0, idx, *pnid;
+
+	// Replace existing if it's there. Extensions are stored in a CSR in an interesting way:
+	//
+	// They are stored as a list under either (most likely) the "official"
+	// NID_ext_req or under NID_ms_ext_req which means everything is stored
+	// under a list in a single "attribute" so we can't use X509_REQ_add1_attr
+	// or similar.
+	//
+	// Instead we have to get the extensions, find and replace the SAN if it's
+	// in there, then *replace* the extensions in the list of attributes. (If
+	// we just try to add it the old ones are found first and don't take
+	// priority)
+
+	has_attrs = X509_REQ_get_attr_count(csr);
+	sk = X509_REQ_get_extensions(csr);
+
+	if (!X509V3_add1_i2d(&sk, target_nid, value, 0, X509V3_ADD_REPLACE))
+		goto error;
+
+	if (X509_REQ_add_extensions(csr, sk) == 0)
+		goto error;
+
+	// Delete the old extensions attribute, so that the one we just added takes priority
```

Comment:
This was part of the mind-bending time I had working this out. So what happens is:

The X509_REQ structure stores a STACK_OF of X509_ATTRIBUTE objects under one of two NIDs (it looks like it's the "official" one is normal, but it can be under an MS named one. 

  ```
  /* It seems several organisations had the same idea of including a list of
 * extensions in a certificate request. There are at least two OIDs that are
 * used and there may be more: so the list is configurable.
 */

static int ext_nid_list[] = { NID_ms_ext_req, NID_ext_req, NID_undef};
  ```

The extensions is stored as a "blob" in DER form (I think. it calls `d2i_ASN1_SET_OF_X509_EXTENSION` on the attribute), so `X509V3_ADD_REPLACE` would replace the subjectAltName in the extensions if there is one, but we need to write back the encoded extensions into the attributes.

Annoying it seems you can add as many attributes under the same NID as you like, but when writing the CSR as PEM (I didn't test any other format) it will only respect the _first_ attribute it finds. So we have to delete the old extensions attribute otherwise the new list is basically ignored.

http://opensource.apple.com//source/OpenSSL096/OpenSSL096-5/openssl/crypto/x509/x509_req.c was the soruce of my reference